File: src/HydrologyFunctions.py
import random
import math
import struct

import Math

# imports and definitions for type hinting
import typing
import DataModel
HydroPrimitive = DataModel.HydroPrimitive
HydrologyNetwork = DataModel.HydrologyNetwork

def selectNode(candidate_nodes: typing.List[HydroPrimitive] , zeta: float) -> HydroPrimitive:
    """Given a list of candidate nodes, this function selects the next node to expand

    The selection is based on Genevaux et al §4.2.1

    :param candidate_nodes: The nodes that can still be expanded
    :type candidate_nodes: list[HydroPrimitive]
    :param zeta: Basically determines how much to prioritize elevation in the selection process (see Genevaux et al §4.2.1)
    :type zeta: int
    :return: The node to expand in the next step
    :rtype: HydroPrimitive
    """
    lowestCandidateZ = min([node.elevation for node in candidate_nodes]) # elevation of lowest candidate
    subselection = [n for n in candidate_nodes if n.elevation < lowestCandidateZ+zeta ] # 
    subselection.sort(key = lambda r : r.priority,reverse = True)
    subsubselection=[node for node in subselection if node.priority == subselection[0].priority]
    if len(subsubselection) > 1:
        subsubselection.sort(key = lambda n : n.elevation)
    
    return subsubselection[0]

class HydrologyParameters:
    """A simple struct that carries the paramaters relevant to expanding the river network

    It's pretty self-explanatory.

    .. todo::
       There must be a better way to do this. In the future it might make sense to
       implement the Factory Pattern to generate the :class:`DataModel.HydrologyNetwork`.
    """
    def __init__(
        self, shore, hydrology, Pa, Pc, maxTries, riverAngleDev, edgeLength, sigma, eta, zeta, riverSlope, slopeRate, candidates
    ):
        self.shore = shore
        self.hydrology = hydrology
        self.Pa = Pa
        self.Pc = Pc
        self.maxTries = maxTries
        self.riverAngleDev = riverAngleDev
        self.edgeLength = edgeLength
        self.sigma = sigma
        self.eta = eta
        self.zeta = zeta
        self.riverSlope = riverSlope
        self.slopeRate = slopeRate
        self.candidates = candidates
    
    def toBinary(self):
        binary =          struct.pack('!f', 0)
        binary = binary + struct.pack('!f', 0)
        binary = binary + struct.pack('!f', self.riverSlope.xSize * self.riverSlope.resolution)
        binary = binary + struct.pack('!f', self.riverSlope.ySize * self.riverSlope.resolution)
        binary = binary + struct.pack('!f', self.Pa)
        binary = binary + struct.pack('!f', self.Pc)
        binary = binary + struct.pack('!f', self.edgeLength)
        binary = binary + struct.pack('!f', self.sigma)
        binary = binary + struct.pack('!f', self.eta)
        binary = binary + struct.pack('!f', self.zeta)
        binary = binary + struct.pack('!f', self.slopeRate)
        binary = binary + struct.pack('!H', self.maxTries)
        binary = binary + struct.pack('!f', self.riverAngleDev)

        binary = binary + struct.pack('!I', self.riverSlope.xSize)
        binary = binary + struct.pack('!I', self.riverSlope.ySize)
        binary = binary + struct.pack('!f', self.shore.resolution)
        binary = binary + self.riverSlope.toBinary()

        # print('ToBinary complete!')

        binary = binary + struct.pack('!I', len(self.candidates))
        for candidate in self.candidates:
            binary = binary + struct.pack('!f', float(candidate.x()))
            binary = binary + struct.pack('!f', float(candidate.y()))
            binary = binary + struct.pack('!I', candidate.priority)
            binary = binary + struct.pack('!Q', candidate.contourIndex)
        
        binary = binary + struct.pack('!Q', len(self.shore.contour))
        for point in self.shore.contour:
            # point = point[0]
            # points in a contour array are y,x
            binary = binary + struct.pack('!Q', point[0])
            binary = binary + struct.pack('!Q', point[1])
        
        return binary

def alpha(node: HydroPrimitive, candidates: typing.List[HydroPrimitive], params: HydrologyParameters):         # alpha, as in the expansion rules in Table 1
    """Alpha node expansion rule

    :param node: The node to expand
    :type node: HydroPrimitive
    :param candidates: The set of candidate nodes
    :type candidates: list[HydroPrimitive]
    :param params: The parameter struct
    :type params: HydrologyParameters
    """
    if node.priority==1:
        ruleBase(node, candidates, params)
    else:
        Pval = random.random();
        if Pval <= params.Pa:
            rulePa(node, candidates, params)
        elif Pval <= params.Pa + params.Pc:
            rulePc(node, candidates, params)
        else:
            rulePs(node, candidates, params)

            
def ruleBase(node: HydroPrimitive, candidates: typing.List[HydroPrimitive], params: HydrologyParameters): #filling
    """Rule 1 from Table 1 in Genevaux et al

    :param node: The node to expand
    :type node: HydroPrimitive
    :param candidates: The set of candidate nodes
    :type candidates: list[HydroPrimitive]
    :param params: The parameter struct
    :type params: HydrologyParameters
    """
    #tao(priority,node)
    for i in range(random.randint(1,5)):
        beta(node, node.priority, candidates, params)
        
        
def rulePc(node: HydroPrimitive, candidates: typing.List[HydroPrimitive], params: HydrologyParameters): #rive growth
    """Rule 2.1 from Table 1 in Genevaux et al

    :param node: The node to expand
    :type node: HydroPrimitive
    :param candidates: The set of candidate nodes
    :type candidates: list[HydroPrimitive]
    :param params: The parameter struct
    :type params: HydrologyParameters
    """
    #tao(priority,node)
    beta(node, node.priority,candidates, params)
    
    
def rulePs(node: HydroPrimitive, candidates: typing.List[HydroPrimitive], params: HydrologyParameters): #symetric junction
    """Rule 2.2 from Table 1 in Genevaux et al

    :param node: The node to expand
    :type node: HydroPrimitive
    :param candidates: The set of candidate nodes
    :type candidates: list[HydroPrimitive]
    :param params: The parameter struct
    :type params: HydrologyParameters
    """
    #tao(priority,node)
    beta(node, node.priority-1, candidates, params)
    beta(node, node.priority-1, candidates, params)

    
def rulePa(node: HydroPrimitive, candidates: typing.List[HydroPrimitive], params: HydrologyParameters): # asymetric junction
    """Rule 2.3 from Table 1 in Genevaux et al

    :param node: The node to expand
    :type node: HydroPrimitive
    :param candidates: The set of candidate nodes
    :type candidates: list[HydroPrimitive]
    :param params: The parameter struct
    :type params: HydrologyParameters
    """
    #tao(priority,node)
    beta(node, node.priority, candidates, params)
    beta(random.randint(1,priority-1),node, params)
    
    
def beta(node: HydroPrimitive, priority: int, candidates: typing.List[HydroPrimitive], params: HydrologyParameters):
    """Instantiates a new node with a given priority

    :param node: The node to expand
    :type node: HydroPrimitive
    :param priority: The priority for the new node
    :type priority: int
    :param candidates: The set of candidate nodes
    :type candidates: list[HydroPrimitive]
    :param params: The parameter struct
    :type params: HydrologyParameters
    """
    point = picknewnodepos(node, params)
    if point is not None:
        slope = params.slopeRate * params.riverSlope[node.x() , node.y()]/255
        newZ = node.elevation + slope * params.edgeLength
        candidates.append(params.hydrology.addNode(point, priority=priority, elevation=newZ, parent=node))
    else:
        tao(node, candidates)

def tao(node: HydroPrimitive, candidates: typing.List[HydroPrimitive]):
    """Removes a node from the set of candidates (rule 3.2 from Table 1 of Genevaux et al)

    :param node: The node to remove
    :type node: HydroPrimitive
    :param candidates: The set (list) of candidates
    :type candidates: list[HydroPrimitive]
    """
    try:
        candidates.remove(node)
    except:
        None
    finally:
        None

def picknewnodepos(parentnode: HydroPrimitive, params: HydrologyParameters) -> typing.Optional[typing.Tuple[float,float]]:
    """Tries to pick a new position for a candidate node

    Tries to avoid the coast and crowding other nodes

    :param parentnode: The node to start from
    :type parentnode: HydroPrimitive
    :param params: The struct holding the relevant parameters
    :type params: HydrologyParameters
    :return: A new position, or None if one could not be found
    :rtype: tuple[float,float] | None
    """
    parentsparent = params.hydrology.downstream(parentnode.id) # parent node of parentnode
    
    angle = None
    if parentsparent is None: # If there is no previous node, Go in a direction perpendicular to the coast
        angle = coastNormal(parentnode, params)
        if angle is None:
            return None
    else:
        # 'angle' is the 'direction' of the river
        angle = math.atan2( # y,x !
            parentnode.y() - parentsparent.y(), # y
            parentnode.x() - parentsparent.x()  # x
        )
    
    newNodePos = None
    # Try maxTries number of times to get a suitable point
    for i in range(params.maxTries):
        # Pick a random new point (generally in the same direction)
        newAngle = angle + random.gauss(0,params.riverAngleDev)
        newNodePos = (
            parentnode.x() + params.edgeLength*math.cos(newAngle),
            parentnode.y() + params.edgeLength*math.sin(newAngle)
        )
        if isAcceptablePosition(newNodePos, params):
            break
        else:
            newNodePos = None
    
    return newNodePos

def coastNormal(node: HydroPrimitive, params: HydrologyParameters) -> float: # Gets the angle that is approximately normal to the coast
    """Gets an angle that is (approximately) perpendicular to the coast

    This is why :class:`DataModel.HydroPrimitive` instances that represent
    a mouth node have a ``contourIndex`` attribute.

    :param node: The node to query
    :type node: HydroPrimitive:
    :param params: The struct holding the relevant parameters
    :type params: HydrologyParameters
    :return: The angle in radians
    :rtype: float
    """
    assert node.contourIndex is not None # assert that this is a mouth node
    p1 = params.shore[node.contourIndex+3]
    p2 = params.shore[node.contourIndex-3]
    theta = math.atan2(p2[1]-p1[1],p2[0]-p1[0])
    return theta - 0.5*math.pi

def isAcceptablePosition(point: typing.Tuple[float,float], params: HydrologyParameters) -> bool:
    """Determines whether or not a point is acceptable as a new node's location

    :param point: The point in question
    :type point: tuple[float,float]
    :param params: The struct holding the relevant parameters
    :type params: HydrologyParameters
    :return: True if the point meets the criteria, False if otherwise
    :rtype: bool
    """
    if point is None:
        return False
    # is the point too close to the seeeeeeeeeeeee?
    if params.shore.distanceToShore(point) < params.eta * params.edgeLength:
        #print(f'Angle too close to the seeee (distance {cv.pointPolygonTest(contour,(point[1],point[0]),True)})')
        return False
    # is the point too close to other nodes?
    for node0,node1 in params.hydrology.edgesWithinRadius(point, 2*params.edgeLength): # Go through each edge
        dist = Math.point_segment_distance( # Distance to the edge (edge is a line segment)
            point[0],point[1],
            node0.x(),node0.y(), # Line segment endpoint 1 (x,y)
            node1.x(),node1.y()  # Line segment endpoint 2 (x,y)
        )
        if dist < params.sigma * params.edgeLength:
            return False
    # otherwise return True
    return True

def classify(node: HydroPrimitive, hydrology: HydrologyNetwork, edgeLength: float):
    """Computes the Rosgen classification for a stretch of river

    :param node: The node to classify
    :type node: HydroPrimitive
    :param hydrology: The hydrological network
    :type hydrology: HydrologyNetwork
    :param edgeLength: The length of each edge between nodes
    :type edgeLength: float

    .. todo::
       This classification is just a slapdash placeholder. A real classification should
       be written in the future.
    """
    # Based on river slope and distance from Gamma
    # TODO: A real classification
    children = hydrology.upstream(node.id)
    for child in children:
        grade = (child.elevation - node.elevation) / edgeLength
        if grade > 0.1:
            child.rosgen = 'A+'
        elif grade >0.04:
            child.rosgen = 'A'
        elif grade > 0.02:
            child.rosgen = ['G','D','B'][random.randint(0,2)];
        elif grade > 0.005:
            child.rosgen = ['C','D','E','F'][random.randint(0,3)];
        else :
            child.rosgen = 'DA'

def getLocalWatershed(node: HydroPrimitive, cells: DataModel.TerrainHoneycomb) -> float:
    """Gets the area of the watershed represented by this particular cell. This is just the cell's area.
    
    :param node: The node of the cell
    :type node: HydroPrimitive
    :param cells: The terrain honeycomb for the terrain
    :type cells: TerrainHoneycomb
    :return: The area of the local watershed in square meters
    :rtype: float
    """
    return cells.cellArea(node)

def getInheritedWatershed(node: HydroPrimitive, hydrology: HydrologyNetwork) -> float:
    """Gets the total area of the watershed.

    This is the area of all the cells that drain into this cell. This includes this cell's area.

    :param node: The node of the cell
    :type node: HydroPrimitive
    :param hydrology: The hydrology network for the terrain
    :type hydrology: HydrologyNetwork
    :return: The area of the inherited watershed in square meters
    :rtype: float
    
    .. note::
        To work properly, the nodes must all have their local watershed areas set. Use :func:`HydrologyFunctions.getLocalWatershed` for this.
    """
    return node.localWatershed + sum([n.inheritedWatershed for n in hydrology.upstream(node.id)])

def getFlow(inheritedWatershed: float) -> float:
    """This estimates the flow of water through a cell.

    This is based on a simple formula specified in §5.1 of Geneveaux et al

    :param inheritedWatershed: The total area of the watershed. Use :func:`HydrologyFunctions.getInheritedWatershed`
    :type inheritedWatershed: float
    :return: The approximate flow in cubic meters per second
    :rtype: float
    """
    return 0.42 * inheritedWatershed**0.69

File: src/hydrologyFunctions.cpp
#include "hydrologyFunctions.hpp"
#include "point.hpp"

#include <random>

#include <opencv2/imgproc.hpp>

#define FLOAT_THRESH 0.1f

class ComparePrimitives
{
  private:
  HydrologyParameters& params;

  public:
  ComparePrimitives(HydrologyParameters& params)
  : params(params)
  {}
  ~ComparePrimitives() {}
  bool operator() (Primitive *a, Primitive *b) {
    return (a->getPriority() > b->getPriority());
  }
};

Primitive selectNode(HydrologyParameters& params) {
  // "find the elevation of the lowest candidate"
  float lowestCandidateZ = params.candidates[0]->getElevation();
  for (size_t i = 0; i < params.candidates.size(); i++)
  {
    if
    (
      params.candidates[i]->getElevation()
      <
      lowestCandidateZ
    )
    {
      lowestCandidateZ = params.candidates[i]->getElevation();
    }
  }

  // consider the subset of admissible nodes made of nodes
  // whose elevations are between that of the lowest candidate,
  // and that elevation plus zeta
  std::vector<Primitive*> subselection;
  for (size_t i = 0; i < params.candidates.size(); i++)
  {
    if
    (
      params.candidates[i]->getElevation()
      <
      (lowestCandidateZ + params.zeta)
    )
    {
      subselection.push_back(params.candidates[i]);
    }
  }

  // sort by priority
  std::sort(
    subselection.begin(),
    subselection.end(),
    ComparePrimitives(params)
  );

  // pick the node with the highest priority and lowest elevation
  size_t idx = 0;
  Primitive lowestElevation = *subselection[idx];
  while
  (
    idx < subselection.size() &&
    subselection[idx]->getPriority() == subselection[0]->getPriority()
  )
  {
    if
    (
      subselection[idx]->getElevation()
      <
      lowestElevation.getElevation()
    )
    {
      lowestElevation = *subselection[idx];
    }
    idx++;
  }

  return lowestElevation;
}

float point_segment_distance(float px, float py, float x1, float y1, float x2, float y2) {
  float dx = x2 - x1;
  float dy = y2 - y1;

  if (abs(dx) < FLOAT_THRESH && abs(dy) < 0)  // the segment's just a point
  {
    return hypotf32(px - x1, py - y1);
  }

  // Calculate the t that minimizes the distance.
  float t = ((px - x1) * dx + (py - y1) * dy) / (dx * dx + dy * dy);

  // See if this represents one of the segment's
  // end points or a point in the middle.
  if (t < 0)
  {
    dx = px - x1;
    dy = py - y1;
  }
  else if (t > 1)
  {
    dx = px - x2;
    dy = py - y2;
  }
  else
  {
    float near_x = x1 + t * dx;
    float near_y = y1 + t * dy;
    dx = px - near_x;
    dy = py - near_y;
  }

  return hypotf32(dx, dy);
}

bool isAcceptablePosition(Point testLoc, float radius, size_t parentID, HydrologyParameters& params) {
  float distToGamma = params.shore.distanceToShore(testLoc.x(), testLoc.y());
  if (distToGamma < params.edgeLength * params.eta)
  { // if the point is too close to the shore, reject it
    return false;
  }
  
  // check all nearby edges
  std::vector<Edge> edges = params.hydrology.queryArea(
    testLoc, radius
  );
  for (Edge edge : edges)
  {
    float dist = point_segment_distance(
      testLoc.x(), testLoc.y(),
      edge.node0->getLoc().x(), edge.node0->getLoc().y(),
      edge.node1->getLoc().x(), edge.node1->getLoc().y()
    );
    if (dist < params.sigma * params.edgeLength)
    { // if the point is too close to any edge, reject it
      return false;
    }
  }
  return true;
}

float coastNormal(Primitive candidate, HydrologyParameters& params) {
  //get two points on the shore that are close to the candidate node
  Point p1(
    params.shore[candidate.getContourIndex()+3].x,
    params.shore[candidate.getContourIndex()+3].y
  );
  Point p2(
    params.shore[candidate.getContourIndex()-3].x,
    params.shore[candidate.getContourIndex()-3].y
  );
  //get an angle that is perpendicular to the shore line
  float theta = atan2(p2.y() - p1.y(), p2.x() - p1.x());
  return theta + M_PI_2f32;
}

LockedPoint::LockedPoint(Point point, AreaLock lock)
: point(point), lock(lock)
{}

void LockedPoint::release() {
  lock.release();
}

Point LockedPoint::getLoc() {
  return point;
}

LockedPoint pickNewNodeLoc(Primitive candidate, HydrologyParameters& params) {
  float angle;
  if (!candidate.hasParent())
  { // if candidate has no parent, then get an angle perpendicular to the coast
    angle = coastNormal(candidate, params);
  }
  else
  {
    // else 'angle' will be the direction of the river
    angle = atan2(
      candidate.getLoc().y() - candidate.getParent()->getLoc().y(),
      candidate.getLoc().x() - candidate.getParent()->getLoc().x()
    );
  }

  float newAngle;
  Point newLoc(-1,-1); //a fake point that will be rejected
  AreaLock lock;
  for (size_t i = 0; i < params.maxTries; i++)
  {
    // get an angle that is somewhat varied from the river's current path
    newAngle = angle + params.distribution(params.generator);
    newLoc = Point(
      candidate.getLoc().x() + cosf32(newAngle) * params.edgeLength,
      candidate.getLoc().y() + sinf32(newAngle) * params.edgeLength
    );
    // lock the area to prevent other threads from modifying it
    lock = params.hydrology.lockArea(newLoc, 2 * params.edgeLength);
    if (isAcceptablePosition(newLoc, 2 * params.edgeLength, candidate.getID(), params))
    {
      break;
    }
    else
    {
      lock.release();
      newLoc = Point(-1,-1); // set the point to a fake point
    }
  }

  return LockedPoint(newLoc, lock);
}

// remove this node from the candidate vector
void tao(Primitive node, HydrologyParameters& params) {
  params.lockCandidateVector();
  size_t candidateIdx = 0;
  for (size_t i = 0; i < params.candidates.size(); i++)
  {
    if (params.candidates[i]->getID() == node.getID())
    {
      candidateIdx = i;
      params.candidates.erase(params.candidates.begin() + candidateIdx);
      break;
    }
  }
  params.unlockCandidateVector();
}

void beta
(Primitive node, int priority, HydrologyParameters& params)
{
  LockedPoint lockedPoint = pickNewNodeLoc(node, params);
  Point newLoc = lockedPoint.getLoc();
  if (newLoc.x() != -1) // if it's something other than the fake point
  {
    float slope = // slope to calculate the new elevation
      params.slopeRate *
      params.riverSlope.get(node.getLoc().x(), node.getLoc().y()) / 255
    ;
    float newZ = node.getElevation() + slope * params.edgeLength;
    params.lockCandidateVector();
      params.candidates.push_back( // add the new node to the candidate vector
        params.hydrology.addRegularNode( // and the hydrology
          newLoc, newZ, priority, node.getID()
        )
      );
    params.unlockCandidateVector();
    lockedPoint.release();
  }
  else
  {
    tao(node, params);
  }
}

void ruleBase(Primitive candidate, HydrologyParameters& params) {
  const int numBranches = 5;
  for (int i = 0; i < numBranches; i++)
  {
    beta(candidate, candidate.getPriority(), params);
  }
}

void pa(Primitive candidate, HydrologyParameters& params) {
  beta(candidate, candidate.getPriority(), params);
  beta(candidate, candidate.getPriority() - 1, params);
}

void pc(Primitive candidate, HydrologyParameters& params) {
  beta(candidate, candidate.getPriority(), params);
}

void ps(Primitive candidate, HydrologyParameters& params) {
  beta(candidate, candidate.getPriority() - 1, params);
  beta(candidate, candidate.getPriority() - 1, params);
}

void alpha(Primitive candidate, HydrologyParameters& params) {
  if (candidate.getPriority() == 1)
  {
    ruleBase(candidate, params);
  }
  else
  {
    float pval = rand();
    if (pval <= params.Pa)
    {
      pa(candidate, params);
    }
    else if (pval <= (params.Pa + params.Pc))
    {
      pc(candidate, params);
    }
    else
    {
      ps(candidate, params);
    }
  }
}